dtm0/clk_src.c

```diff
@@ -0,0 +1,201 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_DTM0
+#include "lib/trace.h"
+
+#include "dtm0/clk_src.h"
+#include "lib/errno.h" /* EINVAL */
+#include "lib/assert.h" /* M0_ASSERT and so on */
+
+enum {
+	/* The hard limit in the now() busy loop.
+	 * The program should panic() if the physical
+	 * clock did not advance for too long (more than
+	 * MAX_RETRIES * ADV_NS nanoseconds).
+	 */
+	CS_PHYS_NOW_MAX_RETRIES = 16,
+
+	/* A soft limit in the now() busy loop.
+	 * This limit should be as small as possible but
+	 * big enough to make the clock advance.
+	 */
+	CS_PHYS_NOW_ADV_NS = 1,
+};
+
+struct m0_dtm0_clk_src_ops {
+	int (*cmp)(const struct m0_dtm0_ts *left,
+		   const struct m0_dtm0_ts *right);
+	int (*now)(struct m0_dtm0_clk_src *cs, struct m0_dtm0_ts *ts);
+	int (*observed)(struct m0_dtm0_clk_src  *cs,
+			const struct m0_dtm0_ts *other_ts);
+};
+
+/* See M0_DTM0_CS_PHYS for details */
+static const struct m0_dtm0_clk_src_ops cs_phys_ops;
+
+M0_INTERNAL int m0_dtm0_clk_src_init(struct m0_dtm0_clk_src *cs,
+				    enum m0_dtm0_cs_types    type)
+{
+	M0_PRE(M0_IN(type, (M0_DTM0_CS_PHYS)));
+	m0_mutex_init(&cs->cs_phys_lock);
+	cs->cs_prev = M0_DTM0_TS_MIN;
+	cs->cs_ops = &cs_phys_ops;
+	return 0;
+}
+
+M0_INTERNAL void m0_dtm0_clk_src_fini(struct m0_dtm0_clk_src *cs)
+{
+	M0_PRE(cs);
+	M0_PRE(M0_IN(cs->cs_ops, (&cs_phys_ops)));
+	m0_mutex_fini(&cs->cs_phys_lock);
+	cs->cs_prev = M0_DTM0_TS_INIT;
+	cs->cs_ops = NULL;
+}
+
+M0_INTERNAL enum m0_dtm0_ts_ord m0_dtm0_ts_cmp(const struct m0_dtm0_clk_src *cs,
+					       const struct m0_dtm0_ts *left,
+					       const struct m0_dtm0_ts *right)
+{
+	M0_PRE(m0_dtm0_ts__invariant(left));
+	M0_PRE(m0_dtm0_ts__invariant(right));
+	return cs->cs_ops->cmp(left, right);
+}
+
+M0_INTERNAL int m0_dtm0_clk_src_now(struct m0_dtm0_clk_src *cs,
+				    struct m0_dtm0_ts      *ts)
+{
+	return cs->cs_ops->now(cs, ts);
+}
+
+M0_INTERNAL int m0_dtm0_clk_src_observed(struct m0_dtm0_clk_src *cs,
+					 const struct m0_dtm0_ts *other_ts)
+{
+	M0_PRE(m0_dtm0_ts__invariant(other_ts));
+	return cs->cs_ops->observed(cs, other_ts);
+}
+
+M0_INTERNAL bool m0_dtm0_ts__invariant(const struct m0_dtm0_ts *ts)
+{
+	const struct m0_dtm0_ts min_ts = M0_DTM0_TS_MIN;
+	const struct m0_dtm0_ts max_ts = M0_DTM0_TS_MAX;
+
+	return _0C(ts->dts_phys >= min_ts.dts_phys) &&
+		_0C(ts->dts_phys <= max_ts.dts_phys);
+}
+
+/* Phys clock implementation */
+
+static enum m0_dtm0_ts_ord cs_phys_cmp(const struct m0_dtm0_ts *left,
+				       const struct m0_dtm0_ts *right)
+{
+	return M0_3WAY(left->dts_phys, right->dts_phys);
+}
+
+static int cs_phys_now(struct m0_dtm0_clk_src *cs, struct m0_dtm0_ts *ts)
```

Comment:
Can something simpler, like

static m0_time_t unique_time(void)
{
	static m0_time_t last = 0;
	m0_time_t        now  = m0_time_now();
	if (now > last) {
		last = now;
	} else {
		last++;
	}
	return last;
}

work? cs_phys_now() should return void.

Reply:
Rewritten, done.